pkg/controllers/node/controller.go

```diff
@@ -45,50 +53,97 @@ func NewController(kubeClient client.Client) *Controller {
 type Controller struct {
 	kubeClient client.Client
 	readiness  *Readiness
+	liveness   *Liveness
+	emptiness  *Emptiness
+	expiration *Expiration
 	finalizer  *Finalizer
 }
 
 // Reconcile executes a reallocation control loop for the resource
 func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
 	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("Node"))
-	// 1. Retrieve node from reconcile request
+	// 1. Retrieve Node, ignore if not provisioned or terminating
 	stored := &v1.Node{}
 	if err := c.kubeClient.Get(ctx, req.NamespacedName, stored); err != nil {
 		if errors.IsNotFound(err) {
 			return reconcile.Result{}, nil
 		}
-		return reconcile.Result{}, err
+		return result.RetryIfError(ctx, err)
 	}
 	if _, ok := stored.Labels[v1alpha3.ProvisionerNameLabelKey]; !ok {
 		return reconcile.Result{}, nil
 	}
+	if !stored.DeletionTimestamp.IsZero() {
+		return reconcile.Result{}, nil
+	}
 
-	// 2. Execute node reconcilers
+	// 2. Retrieve Provisioner
+	provisioner := &v1alpha3.Provisioner{}
+	if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: stored.Labels[v1alpha3.ProvisionerNameLabelKey]}, provisioner); err != nil {
+		return result.RetryIfError(ctx, err)
+	}
+
+	// 3. Execute reconcilers
 	node := stored.DeepCopy()
+	var backoffs []time.Duration
 	var errs error
 	for _, reconciler := range []interface {
-		Reconcile(*v1.Node) error
+		Reconcile(context.Context, *v1alpha3.Provisioner, *v1.Node) (reconcile.Result, error)
 	}{
 		c.readiness,
+		c.liveness,
+		c.expiration,
+		c.emptiness,
```

Comment:
If a node doesn't have a finalizer but will be expired for emptiness/expiration, we could choose to delete this node before we add the finalizer since the expiration and emptiness controllers send a delete request before we patch the node.

In addition, since we're running these in parallel, is it possible that the order in which we execute these controllers change?

Reply:
Semantically, I want the order of these to not matter. I know that in theory, there are some edge cases (i.e, node comes online, instantly ready, we can't reconcile, and emptyttl is 0), but I think they're contrived enough that I'd much rather define an invariant where the order of these reconcilers don't matter.

Reply:
I agree that these cases probably won't happen naturally. I also think our logging is in a good enough state that we would be able to debug these cases if they do happen naturally. From then, I'd love to reconsider this approach if necessary.